plugins/woocommerce/includes/data-stores/abstract-wc-order-data-store-cpt.php

```diff
@@ -715,15 +715,27 @@ protected function update_order_meta_from_object( $order ) {
 
 		foreach ( $order->get_meta_data() as $meta_data ) {
 			if ( isset( $existing_meta_data[ $meta_data->key ] ) ) {
-				// We don't know if the meta is single or array, so we assume it to be array.
+				// We don't know if the meta is single or array, so we assume it to be an array.
 				$meta_value = is_array( $meta_data->value ) ? $meta_data->value : array( $meta_data->value );
+
 				if ( $existing_meta_data[ $meta_data->key ] === $meta_value ) {
 					unset( $existing_meta_data[ $meta_data->key ] );
 					continue;
 				}
 
+				if ( is_array( $existing_meta_data[ $meta_data->key ] ) ) {
+					$key_search = array_search( $meta_data->value, $existing_meta_data[ $meta_data->key ], true );
+					if ( false !== $key_search ) {
+						unset( $existing_meta_data[ $meta_data->key ][ $key_search ] );
+						if ( 0 === count( $existing_meta_data[ $meta_data->key ] ) ) {
+							unset( $existing_meta_data[ $meta_data->key ] );
+						}
+						continue;
+					}
+				}
+
 				unset( $existing_meta_data[ $meta_data->key ] );
-				delete_post_meta( $order->get_id(), $meta_data->key );
+				delete_post_meta( $order->get_id(), $meta_data->key, $meta_data->value );
```

Comment:
You are right, we will never get here, I have removed this code. I think I mean to modify these lines https://github.com/woocommerce/woocommerce/pull/40129/commits/007cf95d9392eba1a980851bad0033943798881b#diff-3bb9bce13a75a81b0fb3401e2655a63fb8fc9d9318cfdf01fb2412d37275b326R751 to prevent some duplicate ops. Previously, it would remove all metadata for the keys with any mismatch, and insert it all again, which is now corrected to only removing metadata that we know is not available in the object.